In [1]:
import requests
import re
from bs4 import BeautifulSoup
import numpy as np
import pandas as pd
import time
from geopy.distance import vincenty

Магический стриппер! (ﾉ◕ヮ◕)ﾉ*:･ﾟ✧

In [4]:
def html_stripper(text):
    return re.sub('<[^<]+?>', '', str(text))

Еще одна магическая штука, которую я нашел на просторах гугла - она как-то работает! Это пригодится нам потом, когда мы будем сортировать массив строк в числовом порядке.

In [5]:
numbers = re.compile(r'(\d+)')
def numericalSort(value):
    parts = numbers.split(value)
    parts[1::2] = map(int, parts[1::2])
    return parts

Поиграв немного с ВАО, я понял, что можно сделать оптимальное разбиение по цене (например, 6.5М рублей), и тогда в каждом запросе будет не более 840 квартир.

In [6]:
urls = ['http://www.cian.ru/cat.php?deal_type=sale&district%5B0%5D={}&engine_version=2&maxprice=6500000&offer_type=flat&p={}&room1=1&room2=1&room3=1&room4=1&room5=1&room6=1',
            'http://www.cian.ru/cat.php?deal_type=sale&district%5B0%5D={}&engine_version=2&minprice=6500001&offer_type=flat&p={}&room1=1&room2=1&room3=1&room4=1&room5=1&room6=1']

In [8]:
start_time = time.time()
links = []
# 56, ..., 71 - номера районов, соответствующих ВАО
for dist in range(56, 72):
    start_time_dist = time.time()
    last_len_links = len(links)
    print('DISTRICT {}'.format(dist))
    for url in urls:
        # Если по запросу выдается, скажем, 16 страниц, то если мы перескочим на 17, нас перенаправит
        # на первую. Переменная is_looped отвечает за то, чтобы мы не зациклились
        is_looped = 0
        page = 1
        while is_looped == 0:
            # тут мы просто собираем информацию обо всех объявлениях на странице
            # (при помощи пристального вглядывания в html-код и профессионального кодинга (͡° ͜ʖ ͡°))
            page_url =  url.format(dist, page)
            search_page = BeautifulSoup(requests.get(page_url).content, 'lxml')
            pages = search_page.findAll('div', attrs = {'class':"pager_pages"})
            pages = re.split('<span>|</span>', str(pages))
            # условие зацикливания
            if page != 1 and ('1' in pages or pages[0] == '[]'):
                print('Total pages: ', page - 1)
                print('Total urls: ', len(links) - last_len_links)
                print('Unique urls: ', len(list(set(links))) - last_len_links)
                is_looped = 1
            # если не зациклились, то достаем все ссылки и переходим к следующей странице
            else:
                flat_urls = search_page.findAll('div', attrs = {'ng-class':"{'serp-item_removed': offer.remove.state, 'serp-item_popup-opened': isPopupOpen}"})
                flat_urls = re.split('http://www.cian.ru/sale/flat/|/" ng-class="', str(flat_urls))
                cnt = 0
                for link in flat_urls:
                    if link.isdigit():
                        links.append(link)
                        cnt += 1
                print('Number of flat urls on the page {}: '.format(page), cnt)
                page += 1
    # нам нужны только уникальные ссылки - как видно из вывода, некоторые объявления повторяются
    links = list(set(links))
    print('Total unique flat urls for district {}: '.format(dist), len(links) - last_len_links)
    print('Time spent on district {}: {} minutes'.format(dist, (time.time() - start_time_dist) / 60))
print('Total time spent: {} munutes'.format((time.time() - start_time) / 60))
print('Total unique flat urls: {}'.format(len(links)))

DISTRICT 56
Number of flat urls on the page 1:  28
Number of flat urls on the page 2:  28
Number of flat urls on the page 3:  28
Number of flat urls on the page 4:  28
Number of flat urls on the page 5:  28
Number of flat urls on the page 6:  28
Number of flat urls on the page 7:  8
Total pages:  7
Total urls:  176
Unique urls:  159
Number of flat urls on the page 1:  28
Number of flat urls on the page 2:  28
Number of flat urls on the page 3:  28
Number of flat urls on the page 4:  28
Number of flat urls on the page 5:  28
Number of flat urls on the page 6:  28
Number of flat urls on the page 7:  28
Number of flat urls on the page 8:  28
Number of flat urls on the page 9:  28
Number of flat urls on the page 10:  28
Number of flat urls on the page 11:  28
Number of flat urls on the page 12:  28
Number of flat urls on the page 13:  28
Number of flat urls on the page 14:  28
Number of flat urls on the page 15:  28
Number of flat urls on the page 16:  28
Number of flat urls on the page 17

Ура, мы получили массив номеров ссылок! Отсортируем его, вспомнив про магическую функцию сравнения в самом начале.

In [9]:
links = sorted(links, key=numericalSort)

In [46]:
len(links)

6994

In [10]:
links[:7]

['1338305', '1535194', '1711530', '1838610', '1986554', '2292702', '2343853']

In [11]:
links[-6:]

['152049291', '152049301', '152049568', '152049833', '152050016', '152050577']

In [17]:
np.savetxt('vao_links.csv', links, delimiter=',', header='link', comments='', fmt='%s')

Теперь займемся функциями, с помощью которых мы сможем доставать интересующую нас информацию со страничек.
Начнем с

> **N – номер квартиры по порядку**

In [18]:
def getN(link):
    return int(link)

(это было слишком просто)
> **Rooms – число комнат**

In [19]:
def getRooms(flat_page):
    rooms = flat_page.find('div', attrs={'class':"object_descr_title"})
    rooms = re.split(' |-', str(rooms))
    rooms = ''.join([i for i in rooms if i.isdigit()])
    # если квартира многокомнатная (6+), то количества комнат в том месте, где мы ищем, не стоит, и
    # у нас вылетает ValueError. Искать это самое число непросто - на разных подобных страничках
    # оно может стоять в разных местах, может быть в тексте, а может вообще не существовать. Поэтому
    # я решил не запариваться и выводить в подобных случаях 6, благо таких случаев на весь ВАО 11
    try:
        return int(rooms)
    except ValueError:
        return 6

> **Price – цена**

В цене на сайте стоят пробельчики, но мы их убираем

In [20]:
def getPrice(flat_page):
    price = flat_page.find('div', attrs={'class':'object_descr_price'})
    price = re.split('<div>|руб|\W', str(price))
    price = ''.join([i for i in price if i.isdigit()][-3:])
    return int(price)

> **Totsp – общая площадь квартиры, кв. м.**

In [21]:
def getTotsp(flat_page):
    totsp = flat_page.find('table', attrs={'class': "object_descr_props flat sale"})
    totsp = html_stripper(totsp).replace('\n', '')
    totsp = re.split('Общая площадь:|\ха0м2', totsp)
    totsp = re.split('\xa0м2', totsp[1])
    # бывает, что некоторых данных на странице нет, как, например, общей площади квартиры.
    # В таких случаях при ValueError мы будем выводить -1 (такие значения в обычном случае все равно
    # нигде не встретятся) и хитро с этим справляться в самом конце
    try:
        return float(totsp[0].replace(',', '.'))
    except ValueError:
        return -1

> **Livesp – жилая площадь квартиры, кв. м.**

(аналогично)

In [22]:
def getLivesp(flat_page):
    livesp = flat_page.find('table', attrs={'class': "object_descr_props flat sale"})
    livesp = html_stripper(livesp).replace('\n', '')
    livesp = re.split('Жилая площадь:|\ха0м2', livesp)
    livesp = re.split('\xa0м2', livesp[1])
    try:
        return float(livesp[0].replace(',', '.'))
    except ValueError:
        return -1

> **Kitsp – площадь кухни, кв. м.**

(аналогично)

In [23]:
def getKitsp(flat_page):
    kitsp = flat_page.find('table', attrs={'class': "object_descr_props flat sale"})
    kitsp = html_stripper(kitsp).replace('\n', '')
    kitsp = re.split('Площадь кухни:|\ха0м2', kitsp)
    kitsp = re.split('\xa0м2', kitsp[1])
    try:
        return float(kitsp[0].replace(',', '.'))
    except ValueError:
        return -1

> **Dist – расстояние от центра в км. (можно по Пифагору)**

In [24]:
def getDist(flat_page):
    # вытаскиваем координаты из карты
    coords = flat_page.find('div', attrs={'class':"map_info_button_extend"}).contents[1]
    coords = re.split('&amp|center=|%2C', str(coords))
    coords_list = []
    for item in coords:
        if item[0].isdigit():
            coords_list.append(item)
    flat_coords = (float(coords_list[0]), float(coords_list[1]))
    # загугленные координаты центра Москвы
    mos_cent_coords = (55.753559, 37.609218)
    try:
        # пользуемся библиотекой geopy
        return vincenty(flat_coords, mos_cent_coords).kilometers
    except ValueError:
        return -1

> **Metrdist – расстояние до метро в минутах**

In [25]:
def getMetrdist(flat_page):
    metrdist = flat_page.find('span', attrs={'class': "object_item_metro_comment"})
    metrdist = html_stripper(metrdist).replace(' ', '')
    metrdist = metrdist.split('\n')
    dists = [int(i) for i in metrdist if i.isdigit()]
    try:
        return int(min(dists))
    except ValueError:
        return -1

> **Walk – 1 – если пешком от метро, 0 – если на транспорте**

In [26]:
def getWalk(flat_page):
    getwalk = flat_page.find('span', attrs={'class': "object_item_metro_comment"})
    getwalk = html_stripper(getwalk).replace(' ', '')
    if 'пешк' in getwalk:
        return 1
    else:
        return 0

> **Brick – 1 – кирпичный/монолит/жб, 0 – другой**

Вообще-то информация о стенах дома может быть в двух разных местах страницы, поэтому я проверяю оба

In [27]:
def getBrick(flat_page):
    getbrick = flat_page.find('table', attrs={'class': "bti__data"})
    getbrick = html_stripper(getbrick).replace('\n', '')
    getbrick = re.split('Материалы стен|Этажность', getbrick)
    if len(getbrick) != 1:
        if 'кирп' in getbrick[1] or 'мон' in getbrick[1] or 'жб' in getbrick[1]\
                               or'ж/б' in getbrick[1] or 'железобетон' in getbrick[1]:
            return 1
        else:
            return 0
    else:
        getbrick = flat_page.find('table', attrs={'class': "object_descr_props flat sale"})
        getbrick = html_stripper(getbrick).replace('\n', '')
        getbrick = re.split('Тип дома:|Тип продажи:', getbrick)
        try:
            if 'кирп' in getbrick[1] or 'мон' in getbrick[1] or 'жб' in getbrick[1]\
                               or'ж/б' in getbrick[1] or 'железобетон' in getbrick[1]:
                return 1
            else:
                return 0
        except ValueError:
            return -1

> **Tel – 1 – если есть, 0 – нет**

In [28]:
def getTel(flat_page):
    gettel = flat_page.find('table', attrs={'class': "object_descr_props flat sale"})
    gettel = html_stripper(gettel).replace('\n', '')
    gettel = re.split('Телефон:|Вид из окна:', gettel)
    if len(gettel) != 3 or 'не' in gettel[1] or '–' in gettel[1]:
        return 0
    else:
        return 1

> **Bal – 1 – есть балкон/лоджия, 0 – нет**

In [29]:
def getBal(flat_page):
    getbal = flat_page.find('table', attrs={'class': "object_descr_props flat sale"})
    getbal = html_stripper(getbal).replace('\n', '')
    getbal = re.split('Балкон:|Лифт:', getbal)
    if len(getbal) != 3 or 'не' in getbal[1] or '–' in getbal[1]:
        return 0
    else:
        return 1

> **Floor – номер этажа, на котором расположена квартира**

In [30]:
def getFloor(flat_page):
    getfloor = flat_page.find('table', attrs={'class': "object_descr_props flat sale"})
    getfloor = html_stripper(getfloor).replace('\n', '')
    getfloor = re.split('Этаж:|\xa0', getfloor)
    try:
        return int(getfloor[1])
    except ValueError:
        return -1

> **Nfloors – всего этажей в доме**

In [31]:
def getNfloors(flat_page):
    getnfloors = flat_page.find('table', attrs={'class': "object_descr_props flat sale"})
    getnfloors = html_stripper(getnfloors).replace('\n', '')
    getnfloors = re.split('/\xa0|Тип дома:', getnfloors)
    try:
        return int(getnfloors[1])
    except ValueError:
        return -1

> **New – 1 – первичный рынок, 0 - вторичка**

In [32]:
def getNew(flat_page):
    getnew = flat_page.find('table', attrs={'class': "object_descr_props flat sale"})
    getnew = html_stripper(getnew).replace('\n', '')
    getnew = re.split('Тип дома:|Высота потолков:', getnew)
    if 'втор' in getnew[1]:
        return 0
    else:
        return 1

In [44]:
#links = pd.read_csv('vao_links.csv')
#links = list(np.array(links['link']).astype(str))

Ссылки есть, функции есть, мы готовы гулять по страничкам!

In [37]:
columns = ['N', 'Rooms', 'Price', 'Totsp', 'Livesp', 'Kitsp', 'Dist', 'Metrdist', 'Walk', 'Brick',
          'Tel', 'Bal', 'Floor', 'Nfloors', 'New']
df_final = pd.DataFrame(columns=columns)

In [47]:
def processing(links):
    n = len(links)
    # количество ошибок (а они встречаются)
    errors = 0
    for link in links[:]:
        flat_page_url = 'http://www.cian.ru/sale/flat/' + link + '/'
        print('\r Processing link {}, {} of {} ({}% completed, total errors: {})'.format(
                flat_page_url, n - len(links) + errors + 1, n,
                round((n - len(links) + errors + 1) / n * 100, 2), errors), end='')
        try:
            flat_page = BeautifulSoup(requests.get(flat_page_url).content, 'lxml')
            df_final.loc[len(df_final)] = [getN(link), getRooms(flat_page), getPrice(flat_page),
                                 getTotsp(flat_page), getLivesp(flat_page), getKitsp(flat_page),
                                 getDist(flat_page), getMetrdist(flat_page), getWalk(flat_page),
                                 getBrick(flat_page), getTel(flat_page), getBal(flat_page),
                                 getFloor(flat_page), getNfloors(flat_page), getNew(flat_page)]
            # если ссылка обработана без ошибок, удаляем ее из исходного массива
            links.remove(link)
        except OSError or MaxRetryError or ConnectionError or ConnectionResetError:
            errors += 1
    if links:
        print('\nSome links were not processed. Trying to process them again...')
        # после процесса обработки в массиве могут остаться только необработанные ссылки
        # (где были ошибки в связи с соединением и т.п.); значит, если он не пуст, они есть, и надо
        # повторить процедуру еще раз
        return processing(links)
    else:
        print('\nAll links were processed')

In [48]:
start_time = time.time()
processing(links)
print('Time spent: {} minutes'.format((time.time() - start_time) / 60))

 Processing link http://www.cian.ru/sale/flat/152050577/, 6994 of 6994 (100.0% completed, total errors: 407)
Some links were not processed. Trying to process them again...
 Processing link http://www.cian.ru/sale/flat/151978992/, 407 of 407 (100.0% completed, total errors: 0)
All links were processed
Time spent: 333.4347388704618 minutes


Отлично, все получилось!

In [50]:
len(df_final)

7062

7062?.. Значит, некоторые ссылки задублировались?

In [61]:
len(df_final['N'].value_counts()[df_final['N'].value_counts() > 1])

68

Да, действительно, ровно 7062 - 6994 = 68 ссылок задублировалось. С этим мы разберемся потом, а пока - к таблице.

In [57]:
df_final.head(16)

N  Rooms        Price  Totsp  Livesp  Kitsp       Dist  Metrdist  \
0   1338305.0    2.0    9000000.0   56.4    35.1    7.0   8.942096      22.0   
1   1535194.0    2.0   52000000.0   92.0    42.0   18.0   5.830754       1.0   
2   1711530.0    2.0    6500000.0   44.5    28.0    6.0  13.741006      13.0   
3   1838610.0    3.0    6600000.0   52.3    36.1    6.0  16.486721      10.0   
4   1986554.0    3.0   13200000.0   80.4    54.8    8.6   7.735389      10.0   
5   2292702.0    3.0   15000000.0   73.5    45.0   10.0  13.470486       9.0   
6   2343853.0    3.0    9700000.0   70.4    48.3    7.0   6.834289      12.0   
7   2464298.0    1.0   14500000.0   58.0    25.0   15.0  11.137224      12.0   
8   2515421.0    5.0   19000000.0  130.0    90.0    8.0  11.272828      17.0   
9   2547927.0    6.0   90000000.0  400.0   300.0   50.0  13.281683       9.0   
10  2621950.0    3.0  105000000.0  148.0    98.0   13.0   6.818786       9.0   
11  2684980.0    3.0   15000000.0   82.0    60.0    8.0   7.801829       2.0   
12  2721689.0    3.0   12100000.0   78.0    48.0   10.0  15.625352       7.0   
13  2793533.0    1.0   20800000.0   77.0    30.0   30.0  11.967088       5.0   
14  2800725.0    3.0    7500000.0   55.6    39.0    6.0  13.451351       7.0   
15  2861831.0    6.0  100000000.0  330.0    -1.0   -1.0  13.281683       9.0   

    Walk  Brick  Tel  Bal  Floor  Nfloors  New  
0    1.0    1.0  1.0  0.0    2.0      5.0  0.0  
1    1.0    1.0  0.0  0.0    2.0     30.0  0.0  
2    1.0    0.0  1.0  0.0    7.0      9.0  0.0  
3    0.0    1.0  1.0  0.0    1.0      5.0  0.0  
4    0.0    1.0  1.0  1.0    7.0      8.0  0.0  
5    1.0    0.0  0.0  1.0   13.0     17.0  0.0  
6    1.0    1.0  1.0  1.0    5.0      6.0  0.0  
7    1.0    1.0  0.0  1.0    8.0     18.0  0.0  
8    1.0    1.0  1.0  0.0    1.0      5.0  0.0  
9    1.0    1.0  1.0  1.0   19.0     19.0  0.0  
10   1.0    0.0  1.0  0.0    4.0     10.0  0.0  
11   0.0    1.0  0.0  1.0    6.0      8.0  0.0  
12   1.0    0.0  1.0  1.0   14.0     17.0  0.0  
13   1.0    0.0  0.0  0.0   20.0     35.0  0.0  
14   1.0    1.0  1.0  1.0    3.0      4.0  0.0  
15   1.0    1.0  1.0  1.0   19.0     20.0  0.0

In [59]:
df_final.tail(5)

N  Rooms       Price  Totsp  Livesp  Kitsp       Dist  \
7057  151977688.0    4.0  15999000.0  108.0    72.0   11.0   8.050940   
7058  151978285.0    1.0   5990000.0   40.0    18.9   10.4  18.302028   
7059  151978628.0    1.0   5300000.0   34.1    18.9    8.5  11.876061   
7060  151978794.0    1.0   4650000.0   32.0    19.0    6.0  12.653934   
7061  151978992.0    3.0  23500000.0   94.0    55.0   14.0  11.123828   

      Metrdist  Walk  Brick  Tel  Bal  Floor  Nfloors  New  
7057      15.0   1.0    1.0  1.0  0.0    2.0      7.0  0.0  
7058      10.0   0.0    0.0  0.0  1.0   19.0     24.0  0.0  
7059      10.0   0.0    1.0  0.0  1.0    3.0      5.0  0.0  
7060      15.0   1.0    0.0  1.0  1.0    4.0      9.0  0.0  
7061       3.0   1.0    1.0  1.0  1.0   15.0     24.0  0.0

Далее становится понятно, почему для пропущенного значения использовалось -1, а не nan: мы переводим большинство столбцов в int (зачем, например, номеру ссылки тип 'float'?..)

In [63]:
df_final[['N', 'Rooms', 'Price', 'Metrdist', 'Walk', 'Brick',
          'Tel', 'Bal', 'Floor', 'Nfloors', 'New']] = \
df_final[['N', 'Rooms', 'Price', 'Metrdist', 'Walk', 'Brick',
          'Tel', 'Bal', 'Floor', 'Nfloors', 'New']].astype(int)
df_final.head(8)

N  Rooms     Price  Totsp  Livesp  Kitsp       Dist  Metrdist  Walk  \
0  1338305      2   9000000   56.4    35.1    7.0   8.942096        22     1   
1  1535194      2  52000000   92.0    42.0   18.0   5.830754         1     1   
2  1711530      2   6500000   44.5    28.0    6.0  13.741006        13     1   
3  1838610      3   6600000   52.3    36.1    6.0  16.486721        10     0   
4  1986554      3  13200000   80.4    54.8    8.6   7.735389        10     0   
5  2292702      3  15000000   73.5    45.0   10.0  13.470486         9     1   
6  2343853      3   9700000   70.4    48.3    7.0   6.834289        12     1   
7  2464298      1  14500000   58.0    25.0   15.0  11.137224        12     1   

   Brick  Tel  Bal  Floor  Nfloors  New  
0      1    1    0      2        5    0  
1      1    0    0      2       30    0  
2      0    1    0      7        9    0  
3      1    1    0      1        5    0  
4      1    1    1      7        8    0  
5      0    0    1     13       17    0  
6      1    1    1      5        6    0  
7      1    0    1      8       18    0

И вот теперь заменяем -1 на nan:

In [64]:
df_final.replace(-1, np.nan, inplace=True)
df_final.head(16)

N  Rooms      Price  Totsp  Livesp  Kitsp       Dist  Metrdist  \
0   1338305      2    9000000   56.4    35.1    7.0   8.942096      22.0   
1   1535194      2   52000000   92.0    42.0   18.0   5.830754       1.0   
2   1711530      2    6500000   44.5    28.0    6.0  13.741006      13.0   
3   1838610      3    6600000   52.3    36.1    6.0  16.486721      10.0   
4   1986554      3   13200000   80.4    54.8    8.6   7.735389      10.0   
5   2292702      3   15000000   73.5    45.0   10.0  13.470486       9.0   
6   2343853      3    9700000   70.4    48.3    7.0   6.834289      12.0   
7   2464298      1   14500000   58.0    25.0   15.0  11.137224      12.0   
8   2515421      5   19000000  130.0    90.0    8.0  11.272828      17.0   
9   2547927      6   90000000  400.0   300.0   50.0  13.281683       9.0   
10  2621950      3  105000000  148.0    98.0   13.0   6.818786       9.0   
11  2684980      3   15000000   82.0    60.0    8.0   7.801829       2.0   
12  2721689      3   12100000   78.0    48.0   10.0  15.625352       7.0   
13  2793533      1   20800000   77.0    30.0   30.0  11.967088       5.0   
14  2800725      3    7500000   55.6    39.0    6.0  13.451351       7.0   
15  2861831      6  100000000  330.0     NaN    NaN  13.281683       9.0   

    Walk  Brick  Tel  Bal  Floor  Nfloors  New  
0      1      1    1    0      2      5.0    0  
1      1      1    0    0      2     30.0    0  
2      1      0    1    0      7      9.0    0  
3      0      1    1    0      1      5.0    0  
4      0      1    1    1      7      8.0    0  
5      1      0    0    1     13     17.0    0  
6      1      1    1    1      5      6.0    0  
7      1      1    0    1      8     18.0    0  
8      1      1    1    0      1      5.0    0  
9      1      1    1    1     19     19.0    0  
10     1      0    1    0      4     10.0    0  
11     0      1    0    1      6      8.0    0  
12     1      0    1    1     14     17.0    0  
13     1      0    0    0     20     35.0    0  
14     1      1    1    1      3      4.0    0  
15     1      1    1    1     19     20.0    0

Ответ готов!

In [65]:
df_final.to_csv('vao_flats.csv')

Итак, в данной части проекта с помощью парсинга сайта cian.ru были получены данные обо всех продающихся в ВАО квартирах. При желании эту процедуру можно проделать для других районов (а также метро и т.п.), выбрав соответствующий критерий для запросов во избежание потери информации (чтобы по каждому запросу выдавалось не больше 840 квартир).